pkg/strategy/funding/strategy.go

```diff
@@ -0,0 +1,266 @@
+
+			if fundingRate >= s.FundingRate.High {
+				s.Notifiability.Notify("%s funding rate %s is too high! threshold %s",
+					s.Symbol,
+					fundingRate.Percentage(),
+					s.FundingRate.High.Percentage(),
+				)
+			} else {
+				if previousIndex != nil {
+					if s.FundingRate.DiffThreshold == 0 {
+						// 0.6%
+						s.FundingRate.DiffThreshold = fixedpoint.NewFromFloat(0.006 * 0.01)
+					}
+
+					diff := fundingRate - previousIndex.LastFundingRate
+					if diff.Abs() > s.FundingRate.DiffThreshold {
+						s.Notifiability.Notify("%s funding rate changed %s, current funding rate %s",
+							s.Symbol,
+							diff.SignedPercentage(),
+							fundingRate.Percentage(),
+						)
+					}
+				}
+			}
+
+			previousIndex = index
+			if fundingRate24HoursLowIndex != nil {
+				if fundingRate24HoursLowIndex.Time.Before(time.Now().Add(24 * time.Hour)) {
+					fundingRate24HoursLowIndex = index
+				}
+				if fundingRate < fundingRate24HoursLowIndex.LastFundingRate {
+					fundingRate24HoursLowIndex = index
+				}
+			} else {
+				fundingRate24HoursLowIndex = index
+			}
+		}
+	}
+}
+
+func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) error {
+	standardIndicatorSet, ok := session.StandardIndicatorSet(s.Symbol)
+	if !ok {
+		return fmt.Errorf("standardIndicatorSet is nil, symbol %s", s.Symbol)
+	}
+
+	binanceExchange, ok := session.Exchange.(*binance.Exchange)
+	if !ok {
+		log.Error("exchange does not support funding rate api")
+	}
+	binanceExchange.UseFutures()
+	//if s.FundingRate != nil {
+	//	go s.listenToFundingRate(ctx, binanceExchange)
+	//}
+
+	session.MarketDataStream.OnKLineClosed(func(kline types.KLine) {
+		// skip k-lines from other symbols
+		if kline.Symbol != s.Symbol {
+			return
+		}
+		log.Infof(s.Symbol)
+		for _, detection := range s.SupportDetection {
+			if kline.Interval != detection.Interval {
+				continue
+			}
+
+			closePriceF := kline.GetClose()
+			closePrice := fixedpoint.NewFromFloat(closePriceF)
+
+			var ma types.Float64Indicator
+
+			switch strings.ToLower(detection.MovingAverageType) {
+			case "sma":
+				ma = standardIndicatorSet.SMA(types.IntervalWindow{
+					Interval: detection.MovingAverageInterval,
+					Window:   detection.MovingAverageWindow,
+				})
+			case "ema", "ewma":
+				ma = standardIndicatorSet.EWMA(types.IntervalWindow{
+					Interval: detection.MovingAverageInterval,
+					Window:   detection.MovingAverageWindow,
+				})
+			default:
+				ma = standardIndicatorSet.EWMA(types.IntervalWindow{
+					Interval: detection.MovingAverageInterval,
+					Window:   detection.MovingAverageWindow,
+				})
+			}
+
+			var lastMA = ma.Last()
+
+			// skip if the closed price is under the moving average
+			if closePrice.Float64() < lastMA {
+				log.Infof("skip %s support closed price %f > last ma %f", s.Symbol, closePrice.Float64(), lastMA)
+				return
+			}
+
+			premiumIndex, err := binanceExchange.QueryPremiumIndex(ctx, s.Symbol)
```

Comment:
add QueryPremiumIndex method to the FuturesExchange interface, you can assert the exchange from the beginning, and check the error outside this closure